compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp

```diff
@@ -0,0 +1,227 @@
+  : _input_data_path(input_data_path), _qerror_ratio(qerror_ratio), _qerror(0.f),
+    _quantizer(std::make_unique<Quantizer>("uint8", "uint8"))
+{
+}
+
+BisectionSolver::Options *BisectionSolver::options(void)
+{
+  if (_options == nullptr)
+  {
+    _options = std::make_unique<BisectionOptionsImpl>();
+  }
+
+  return _options.get();
+}
+
+float BisectionSolver::evaluate(const DatasetEvaluator &evaluator, const std::string &flt_path,
+                                const std::string &def_quant, LayerParams &layers)
+{
+  auto model = read_module(flt_path);
+  // get fake quantized model for evaluation
+  if (!_quantizer->fake_quantize(model.get(), def_quant, layers))
+  {
+    throw std::runtime_error("Failed to produce fake-quantized model.");
+  }
+
+  return evaluator.evaluate(model.get());
+}
+
+std::unique_ptr<luci::Module> BisectionSolver::run(const std::string &module_path)
+{
+  LOGGER(l);
+
+  auto module = read_module(module_path);
+
+  float min_depth = 0.f;
+  float max_depth = 0.f;
+  NodeDepthType nodes_depth;
+  if (compute_depth(module.get(), nodes_depth, min_depth, max_depth) !=
+      ParameterizerResult::SUCCESS)
+  {
+    std::cerr << "Invalid graph for bisectioning" << std::endl;
+    return nullptr;
+  }
+
+  std::unique_ptr<MAEMetric> metric = std::make_unique<MAEMetric>();
+  DatasetEvaluator evaluator(module.get(), _input_data_path, *metric.get());
+
+  // const auto &ref_output = compute_outputs(module.get(), _input_data_path);
+  LayerParams layer_params;
+  float int16_qerror = evaluate(evaluator, module_path, "int16", layer_params);
+  VERBOSE(l, 0) << "int16_quantization_error " << int16_qerror << std::endl;
+
+  float int8_qerror = evaluate(evaluator, module_path, "uint8", layer_params);
+  VERBOSE(l, 0) << "int8_quantization_error " << int8_qerror << std::endl;
+
+  _qerror = int16_qerror + _qerror_ratio * std::fabs(int8_qerror - int16_qerror);
+  VERBOSE(l, 0) << "target quantization error " << _qerror << std::endl;
+
+  if (int16_qerror > _qerror)
+  {
+    std::cerr << "ERROR: minimal targeted is an error of Q16 quantized model" << std::endl;
+    return nullptr;
+  }
+  if (int8_qerror < _qerror)
+  {
+    // no need for bisectioning just return Q8 model
+    if (!_quantizer->quantize(module.get(), "uint8", layer_params))
+    {
+      std::cerr << "ERROR: failed to quantize model" << std::endl;
+      return nullptr;
+    }
+  }
+
+  int last_depth = -1;
+  float best_depth = -1;
+  LayerParams best_params;
+  auto graph = module->graph(0);
+  auto active_nodes = loco::active_nodes(loco::output_nodes(graph));
+  // input and output nodes are not valid for quantization, so let's remove them
+  for (auto node : loco::input_nodes(graph))
+  {
+    active_nodes.erase(node);
+  }
+  for (auto node : loco::output_nodes(graph))
+  {
+    active_nodes.erase(node);
+  }
+
+  // let's decide whether nodes at input are more suspectible to be quantized into Q16, than at
+  // output
+  bool int16_at_input = true;
```

Comment:
Could you rename the variable? It is a bit hard to understand.

```suggestion
  bool int16_front = true;
```

Reply:
ok